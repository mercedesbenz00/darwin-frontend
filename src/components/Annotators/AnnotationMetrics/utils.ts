import moment from 'moment'

import { DatasetAnnotationReportPayload, MembershipPayload } from '@/store/types'

import {
  AdaptedDataPoint,
  AggregatedDataPoint,
  AnnotationDataGranularity,
  AnnotationDimensionData,
  AnnotationDataRange
} from './types'

/* builds a moment timestamp formatted a string matching those returned from backend */
export const buildFrom = (range: AnnotationDataRange) =>
  moment
    .utc()
    .subtract(1, range as moment.unitOfTime.DurationConstructor)
    .startOf('hour')
    .format('YYYY-MM-DDTHH:00:00.000000')

/**
 * Takes aggregated source report data and addaptes it for rendering on the line chart.
 *
 * This is done by computing necessary data points.
 */
export const adaptReportData = (r: AggregatedDataPoint): AdaptedDataPoint => ({
  annotationTime: r.annotation_time,
  avgTimePerAnnotation: r.annotations_created > 0
    ? r.annotation_time / r.annotations_created
    : null,
  date: r.date,
  userId: r.user_id,
  imagesAnnotated: r.images_annotated,
  annotationsCreated: r.annotations_created,
  reviewPassRate: r.images_approved > 0
    ? r.images_approved / (r.images_approved + r.images_rejected) * 100
    : null
})

/**
 * Returns a default adapted data point for a team member
 */
export const defaultData = (member: MembershipPayload, date: string): AggregatedDataPoint => ({
  date,
  user_id: member.user_id,
  annotation_time: 0,
  annotations_created: 0,
  annotations_approved: 0,
  images_annotated: 0,
  images_approved: 0,
  images_rejected: 0
})

export const getDimensionsData = (data: AdaptedDataPoint[]): AnnotationDimensionData => {
  const annotationTime = data.reduce((acc, r) => acc + r.annotationTime, 0)
  const annotationsCreated = data.reduce((acc, r) => acc + r.annotationsCreated, 0)
  const avgTimePerAnnotation = annotationsCreated === 0 ? null : annotationTime / annotationsCreated
  const imagesAnnotated = data.reduce((acc, r) => acc + r.imagesAnnotated, 0)

  const totalPassRate = data.reduce((acc, r) => acc + (r.reviewPassRate || 0), 0)
  const periodsWithPassRate = data.filter(r => r.reviewPassRate !== null).length
  const reviewPassRate = totalPassRate / periodsWithPassRate

  return {
    annotationTime,
    annotationsCreated,
    avgTimePerAnnotation,
    imagesAnnotated,
    reviewPassRate
  }
}

/**
 * Expands spars data points so each possible date in the series is covered.
 *
 * This is done by adding a default data point for any missing date
 *
 */
export const expandData = (
  member: MembershipPayload,
  data: AggregatedDataPoint[],
  dates: string[]
): AggregatedDataPoint[] =>
  dates.map(date => data.find(d => d.date === date) || defaultData(member, date))

export const aggregateByDate =
  (date: string, annotationReport: DatasetAnnotationReportPayload[]): AggregatedDataPoint => {
    const initial = {
      annotation_time: 0,
      annotations_created: 0,
      annotations_approved: 0,
      images_annotated: 0,
      images_approved: 0,
      images_rejected: 0
    }

    const reduced = annotationReport.filter(r => r.date === date).reduce((acc, r) => {
      acc.annotation_time += r.annotation_time
      acc.annotations_created += r.annotations_created
      acc.annotations_approved += r.annotations_approved
      acc.images_annotated += r.images_annotated
      acc.images_approved += r.images_approved
      acc.images_rejected += r.images_rejected
      return acc
    }, initial)

    return { ...reduced, date, user_id: null }
  }

export const dateRange =
  (fromDate: string, granularity: AnnotationDataGranularity, format: string) => {
    const from = moment.utc(fromDate)
    const now = moment.utc()
    const dates: string[] = []

    const step = granularity as moment.unitOfTime.DurationConstructor

    const diff = now.diff(from, step)

    for (let i = 0; i <= diff; i++) {
      dates.push(from.format(format))
      from.add(1, step)
    }

    return dates
  }

export const chartFormat = (granularity: AnnotationDataGranularity): string =>
  granularity === 'hour'
    ? 'YYYY-MM-DDTHH:00:00.000000'
    : 'YYYY-MM-DD'

/**
 * selectable data ranges for which we define a "from" parameter when fetching data
 */
const DEFINED_RANGES: AnnotationDataRange[] = ['month', 'week', 'day']

export const granularityFromRange = (range: AnnotationDataRange) => {
  if (range === 'day') { return 'hour' }
  if (range === 'week') { return 'hour' }
  return 'day'
}

/**
 * Computes a "from" param based on the specified range
 *
 * ```
 * 'total' -> null
 * 'month' -> start of hour a month ago
 * 'week' -> start of hour a week ago
 * 'day' -> start of hour a day ago
 * ```
 */
export const fromFromRange = (range: AnnotationDataRange) =>
  DEFINED_RANGES.includes(range) ? buildFrom(range) : null
