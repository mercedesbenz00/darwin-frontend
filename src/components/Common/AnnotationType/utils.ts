import { AnnotationTypeName } from '@/store/types'

/**
 * List of non-meta annotation types. These are basically all types except
 * `measures` and `auto_annotate`.
 *
 * This list is important because these types usually get rendered and are in
 * some way visible to the user, while the remaining two are internal.
 */
export const SORTED_NON_META_TYPES: AnnotationTypeName[] = [
  'attributes',
  'auto_annotate',
  'bounding_box',
  'cuboid',
  'directional_vector',
  'ellipse',
  'graph',
  'inference',
  'instance_id',
  'keypoint',
  'line',
  'link',
  'measures',
  'polygon',
  'polyline',
  'skeleton',
  'string',
  'table',
  'tag',
  'text'
]

export const annotationTypeNameValidator = (value: string): boolean =>
  SORTED_NON_META_TYPES.some(t => t === value)
