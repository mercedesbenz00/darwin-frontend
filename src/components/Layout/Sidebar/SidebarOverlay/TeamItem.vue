<template>
  <div
    role="button"
    class="team-item"
    :class="isCurrent ? 'team-item--current' : 'team-item--other'"
    @click="$emit('click')"
  >
    <team-avatar
      class="team-item__avatar"
      :team="team"
    />
    <div class="team-item__team">
      <div class="team-item__team__name">
        {{ team.name }}
      </div>
      <partner-pill
        v-if="isPartner"
        class="team-item__team__pill"
      />
      <div
        v-if="isCurrent"
        class="team-item__team__members"
      >
        {{ memberCount }} team members
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'
import { Getter, State } from 'vuex-class'

import TeamAvatar from '@/components/Common/Avatar/V1/TeamAvatar.vue'
import PartnerPill from '@/components/PartnerTeam/PartnerPill.vue'
import { MembershipPayload, TeamPayload } from '@/store/types'

@Component({ name: 'team-item', components: { TeamAvatar, PartnerPill } })
export default class TeamItem extends Vue {
  @Prop({ required: true, type: Object as () => TeamPayload })
  team!: TeamPayload

  @Getter('membershipsForTeam', { namespace: 'team' })
  memberships!: (team: TeamPayload) => MembershipPayload[]

  @State(state => state.team.currentTeam)
  currentTeam!: TeamPayload | null

  get isCurrent (): boolean {
    return this.team.id === (!!this.currentTeam && this.currentTeam.id)
  }

  get memberCount (): number | null {
    return this.isCurrent
      ? this.memberships(this.team).length
      : null
  }

  get isPartner (): boolean {
    return this.team.managed_status === 'partner'
  }
}
</script>

<style lang="scss" scoped>
.team-item:hover {
  background: #CCF4F0;
  cursor: pointer;
}

.team-item {
  @include row--center;
  column-gap: 10px;
  align-items: center;
  padding: 10px 16px;
}

.team-item__avatar {
  width: 40px;
  height: 40px;
  border-radius: 50%;
}

.team-item__avatar :deep(.avatar__icon) {
  border-radius: 50%;
}

.team-item__team {
  display: grid;
  grid-auto-flow: row;
  grid-template-columns: 1fr;
  width: calc(100% - 54px);
  justify-content: start;
}

.team-item__team__name {
  @include ellipsis(2, md-1);
  word-break: break-all;
  font-family: $fontFamilyDefault;
  color: $colorSecondaryDark1;

  grid-row: 1 / 2;
}

.team-item__team .team-item__team__pill {
  grid-row: 1 / 2;

  align-self: start;
  justify-self: end;
}

.team-item--other .team-item__team__name {
  @include ellipsis(1, md-1);
  font-family: $fontFamilyDefault;
  color: $colorSecondaryDark1;
}

.team-item__team__members {
  @include typography(md-1, headlines);
  color: $colorSecondaryLight;
  grid-row: 2 / 3;
}
</style>
