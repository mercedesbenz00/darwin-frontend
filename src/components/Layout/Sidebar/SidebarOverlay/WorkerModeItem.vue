<template>
  <div
    :class="{ 'worker-mode-item': true, 'worker-mode-item--enabled': enabled }"
    @click.stop="toggleWorkerMode"
  >
    <active-annotators-icon
      class="worker-mode-item__icon"
      v-if="enabled"
    />
    <annotators-icon
      class="worker-mode-item__icon"
      v-else
    />
    <div class="worker-mode-item__text">
      Worker Mode
    </div>
    <toggle :value="enabled" />
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import { State } from 'vuex-class'

import { ActiveAnnotatorsIcon, AnnotatorsIcon } from '@/assets/icons/V1/sidebar'
import Toggle from '@/components/Common/Toggle/V1/Toggle.vue'
import { RootState } from '@/store/types'

@Component({
  name: 'worker-mode-item',
  components: { ActiveAnnotatorsIcon, AnnotatorsIcon, Toggle }
})
export default class WorkerModeItem extends Vue {
  @State((state: RootState) => state.ui.workerMode)
  readonly enabled!: boolean

  toggleWorkerMode (): void {
    this.$store.dispatch('ui/toggleWorkerMode')
  }
}
</script>

<style lang="scss" scoped>
.worker-mode-item {
  @include row;
  align-items: center;
  padding: 22px 16px;

  &__icon {
    width: 16px;
    height: 16px;
    margin-right: 8px;
    color: $colorAliceNight;
  }

  &__text {
    @include typography(md-1, headlines);
    flex: 1 0 0;
    margin-right: 8px;
    color: $colorSecondaryLight;
  }

  &--enabled &__text {
    color: $colorFeatherLight;
  }

}
</style>
