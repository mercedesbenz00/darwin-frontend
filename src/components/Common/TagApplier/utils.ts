import { AnnotationClassPayload } from '@/store/types'
import { rgbaStringToHSLAString } from '@/utils'

export const tagBrightenedColor = (color: string): string =>
  rgbaStringToHSLAString(color, { l: 70 })

export const matchClassByName = (
  classes: AnnotationClassPayload[],
  text: string
): AnnotationClassPayload | undefined =>
  classes.find(c => c.name === text)
