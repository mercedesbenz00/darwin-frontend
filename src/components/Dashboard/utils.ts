import { MetricPayload } from '@/store/types'
import { RGBA, rgbaString } from '@/utils'

import { GradientLineChartDataset } from './types'

export const metricChartData = (metric: MetricPayload): GradientLineChartDataset[] => {
  const color: RGBA = { r: 0, g: 217, b: 201, a: 1.0 }
  const gradient = { 0: rgbaString(color, 0.3), 0.5: 'rgba(255, 255, 255, 0)' }
  const fillColor = rgbaString(color, 1)

  return [{
    data: metric.data,
    label: metric.name,
    backgroundColor: { gradient: gradient },
    borderColor: fillColor,
    borderWidth: 1,
    pointRadius: 0,
    pointBackgroundColor: fillColor,
    pointBorderWidth: 0
  }]
}
