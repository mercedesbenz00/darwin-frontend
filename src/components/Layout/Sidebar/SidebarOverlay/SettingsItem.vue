<template>
  <router-link
    class="settings-item"
    :to="path"
  >
    <icon-duotone-settings class="settings-item__icon" />
    <div class="settings-item__text">
      {{ label }}
    </div>
  </router-link>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'

import { IconDuotoneSettings } from '@/assets/icons/V2/Duotone'
import { resolveSettingsRoutePath } from '@/utils/router'

@Component({
  name: 'settings-item', components: { IconDuotoneSettings }
})
export default class SettingsItem extends Vue {
  get path (): string {
    const settingsTab = this.$can('manage_customer') ? 'plans' : 'profile'
    return resolveSettingsRoutePath(this.$router, this.$route, settingsTab)
  }

  get label (): string {
    return this.$can('manage_customer') ? 'Settings & Billing' : 'Settings'
  }
}
</script>

<style lang="scss" scoped>
.settings-item {
  @include row;
  align-items: center;
  padding: 22px 16px;
}

.settings-item__icon {
  width: 16px;
  height: 16px;
  margin-right: 8px;
  color: $colorAliceNight;
}

.settings-item__text {
  @include typography(md-1, headlines);
  color: $colorSecondaryLight;
}
</style>
