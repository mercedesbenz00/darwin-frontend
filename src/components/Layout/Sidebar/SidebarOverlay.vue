<template>
  <div class="overlay">
    <div
      :style="{ '--max-team-amount': maxTeamAmount, '--el-height': elHeight }"
      class="overlay-wrapper"
    >
      <team-item
        :team="currentTeam"
        class="overlay__team overlay__team--current"
        @click="showSettings"
      />
      <team-item
        v-for="client in currentTeam.clients"
        :key="`client-${client.id}`"
        :team="client"
        class="overlay__team overlay__team--current-client"
        @click="selectTeam(client)"
      />
      <template v-for="team in otherTeams">
        <team-item
          :key="team.id"
          :team="team"
          class="overlay__team overlay__team--other"
          @click="selectTeam(team)"
        />
        <team-item
          v-for="client in team.clients"
          :key="`client-${client.id}`"
          :team="client"
          class="overlay__team overlay__team--other-client"
          @click="selectTeam(client)"
        />
      </template>
    </div>
    <div class="overlay__actions">
      <settings-item class="overlay__settings" />
      <feature-guard feature="WORKER_MODE_TOGGLE">
        <worker-mode-item class="overlay__worker-mode" />
      </feature-guard>
      <div
        class="overlay__logout"
        @click="logout"
      >
        <power-icon class="overlay__logout__icon" />
        <div class="overlay__logout__text">
          Log Out
        </div>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import { State } from 'vuex-class'

import { PowerIcon } from '@/assets/icons/V1'
import Avatar from '@/components/Common/Avatar/V1/Avatar.vue'
import FeatureGuard from '@/components/Common/FeatureGuard.vue'
import { TeamPayload } from '@/store/types'
import { notifyErrorByCode } from '@/utils'
import { ErrorCodes } from '@/utils/error/errors'

import SettingsItem from './SidebarOverlay/SettingsItem.vue'
import TeamItem from './SidebarOverlay/TeamItem.vue'
import WorkerModeItem from './SidebarOverlay/WorkerModeItem.vue'

@Component({
  name: 'sidebar-overlay',
  components: { Avatar, FeatureGuard, PowerIcon, SettingsItem, TeamItem, WorkerModeItem }
})

export default class SidebarOverlay extends Vue {
  readonly maxTeamAmount: number = 5
  readonly elHeight: string = '61px'

  @State(state => state.team.currentTeam)
  currentTeam!: TeamPayload

  @State(state => state.team.teams)
  teams!: TeamPayload[]

  get otherTeams (): TeamPayload[] {
    return this.teams.filter(team => team.id !== this.currentTeam.id)
  }

  get teamAmount (): number {
    let length = this.teams.length + this.currentTeam.clients.length
    this.otherTeams.forEach(team => {
      length += team.clients.length
    })

    return length
  }

  showSettings (): void {
    this.$store.dispatch('ui/showSettingsDialog', { tab: 'team' })
  }

  async selectTeam (team: TeamPayload): Promise<void> {
    if (team.disabled) {
      return notifyErrorByCode(this.$store, ErrorCodes.TEAM_DISABLED)
    }

    const { error } = await this.$store.dispatch('auth/selectTeam', { team_id: team.id })
    if (error) { return this.$store.dispatch('toast/warning', { content: error.message }) }

    // hack forces vue router to rerun beforeEach if the user was already at `/`
    this.$router.push('/hack')
    this.$router.replace('/')
  }

  // logout

  logout (): void {
    this.$store.dispatch('auth/logout')
  }
}
</script>

<style lang="scss" scoped>
$borderColor: rgba(203, 209, 222, 0.3);

.overlay {
  background: rgba(203, 209, 222, 0.3);
  position: absolute;
  left: 10px;
  top: 65px;
  width: 260px;
  height: auto;
  border-radius: 10px;
  background: $colorWhite;
  box-shadow: 0px 10px 20px rgba(20, 5, 60, 0.1), 0px 4px 4px rgba(20, 5, 60, 0.2);
  overflow: auto;
  cursor: pointer;
  z-index: 10;
}

.overlay-wrapper {
  @include scrollbar;
  display: block;
  overflow-y: scroll;
  overflow-x: hidden;

  min-height: var(--el-height);
  height: calc((var(--el-height) * var(--max-team-amount)) + (var(--el-height) / 2));
  max-height: 50vh;
  z-index: 1;
}

.overlay__actions {
  background: $colorWhite;
  z-index: 2;
  border-top: 1px solid $borderColor;
}

.overlay__team,
.overlay__settings,
.overlay__worker-mode {
  border-bottom: 1px solid $borderColor;
}

.overlay__team--current-client,
.overlay__team--other-client {
  padding: 10px 28px;
  position: relative;
}

.overlay__team--current-client:before,
.overlay__team--other-client:before {
  content: "";
  display: block;
  width: 3px;
  height: calc(100% - 20px);
  position: absolute;
  left: 16px;
  background: $borderColor;
}

.overlay__logout {
  @include row;
  align-items: center;
  padding: 22px 16px;
}

.overlay__logout__icon {
  width: 16px;
  height: 16px;
  margin-right: 8px;
}

.overlay__logout__text {
  @include typography(md-1, headlines);
  color: $colorSecondaryLight;
}
</style>
