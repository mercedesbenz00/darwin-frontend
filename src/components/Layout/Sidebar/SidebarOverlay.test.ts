import { createLocalVue, shallowMount, Stubs, Wrapper } from '@vue/test-utils'
import flushPromises from 'flush-promises'
import Vuex from 'vuex'

import createTestStore from 'test/unit/createTestStore'
import { buildTeamPayload } from 'test/unit/factories'

import SidebarOverlay from '@/components/Layout/Sidebar/SidebarOverlay.vue'
import clickOutsideDirective from '@/directives/click-outside'
import { errorsByCode } from '@/utils'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.directive('click-outside', clickOutsideDirective)

let store: ReturnType<typeof createTestStore>
let mocks: {
  $router: {
    push: Function,
    replace: Function
  }
}
const stubs: Stubs = ['router-link']

const v7 = buildTeamPayload({
  id: 1,
  name: 'V7',
  image: { id: 1, url: 'foo', thumbnail_url: 'bar' }
})

const otherTeam = buildTeamPayload({
  id: 2,
  name: 'Other Team',
  image: { id: 2, url: 'foo1', thumbnail_url: 'bar1' }
})

const clientTeam1 = buildTeamPayload({ id: 4, name: 'Client 1', managed_status: 'client' })
const clientTeam2 = buildTeamPayload({ id: 5, name: 'Client 2', managed_status: 'client' })

const partnerTeam = buildTeamPayload({
  id: 3,
  name: 'Partner',
  managed_status: 'partner',
  clients: [clientTeam1, clientTeam2]
})

class TeamItemModel {
  wrapper: Wrapper<Vue>

  constructor (wrapper: Wrapper<Vue>) {
    this.wrapper = wrapper
  }

  click (): Vue {
    return this.wrapper.vm.$emit('click')
  }

  get key (): string {
    return this.wrapper.vm.$vnode.key!.toString()
  }
}

class Model {
  wrapper: Wrapper<Vue>

  constructor (wrapper: Wrapper<Vue>) {
    this.wrapper = wrapper
  }

  get currentTeam (): TeamItemModel {
    return new TeamItemModel(this.wrapper.find('.overlay__team--current'))
  }

  get currentClients (): TeamItemModel[] {
    return this
      .wrapper.findAll('.overlay__team--current-client')
      .wrappers.map(wrapper => new TeamItemModel(wrapper))
  }

  get otherTeams (): TeamItemModel[] {
    return this
      .wrapper.findAll('.overlay__team--other')
      .wrappers.map(wrapper => new TeamItemModel(wrapper))
  }

  get otherClients (): TeamItemModel[] {
    return this
      .wrapper.findAll('.overlay__team--other-client')
      .wrappers.map(wrapper => new TeamItemModel(wrapper))
  }
}

beforeEach(() => {
  store = createTestStore()
  store.commit('auth/SET_AUTHENTICATED', true)
  store.commit('team/SET_TEAMS', [v7, otherTeam, partnerTeam])
  store.commit('team/SET_CURRENT_TEAM', v7)

  mocks = {
    $router: {
      push: jest.fn(),
      replace: jest.fn()
    }
  }
})

it('matches snapshot', () => {
  const wrapper = shallowMount(SidebarOverlay, { localVue, mocks, store, stubs })
  expect(wrapper).toMatchSnapshot()
})

it('opens settings dialog when current team is clicked', async () => {
  const wrapper = shallowMount(SidebarOverlay, { localVue, store, stubs, mocks })
  const model = new Model(wrapper)
  await model.currentTeam.click()
  expect(store.dispatch).toBeCalledWith('ui/showSettingsDialog', { tab: 'team' })
})

describe('if other team is clicked', () => {
  it('dispatches store action to select other team', async () => {
    const wrapper = shallowMount(SidebarOverlay, { localVue, store, stubs, mocks })
    const model = new Model(wrapper)
    await model.otherTeams[0].click()
    expect(store.dispatch).toBeCalledWith('auth/selectTeam', { team_id: otherTeam.id })
    await flushPromises()
    expect(mocks.$router.push).toBeCalledWith('/hack')
    expect(mocks.$router.replace).toBeCalledWith('/')
  })

  it('toasts error when other team is disabled', async () => {
    const otherTeam = buildTeamPayload({ id: 2, disabled: true })
    store.commit('team/SET_TEAMS', [v7, otherTeam])

    const wrapper = shallowMount(SidebarOverlay, { localVue, store, stubs, mocks })
    const model = new Model(wrapper)
    await model.otherTeams[0].click()
    expect(store.dispatch).toBeCalledWith('toast/warning', { content: errorsByCode.TEAM_DISABLED })
    await flushPromises()
    expect(mocks.$router.push).not.toBeCalledWith('/hack')
    expect(mocks.$router.replace).not.toBeCalledWith('/')
  })

  it('toasts error if vuex action fails with error', async () => {
    jest.spyOn(store, 'dispatch').mockResolvedValue({ error: { message: 'foo' } })
    const wrapper = shallowMount(SidebarOverlay, { localVue, store, stubs, mocks })
    const model = new Model(wrapper)
    await model.otherTeams[0].click()
    await flushPromises()
    expect(store.dispatch).toBeCalledWith('toast/warning', { content: 'foo' })
    expect(mocks.$router.push).not.toBeCalledWith('/hack')
    expect(mocks.$router.replace).not.toBeCalledWith('/')
  })
})

it('renders clients on unselected partner team', () => {
  const wrapper = shallowMount(SidebarOverlay, { localVue, store, stubs, mocks })
  const model = new Model(wrapper)
  expect(model.otherClients.length).toEqual(2)
  expect(model.otherClients[0].key.startsWith('client')).toBe(true)
})

it('assigns prefixed key to selected team clients', () => {
  const wrapper = shallowMount(SidebarOverlay, { localVue, store, stubs, mocks })
  const model = new Model(wrapper)
  expect(model.otherClients[0].key.startsWith('client')).toBe(true)
})

it('renders clients on current partner team', () => {
  store.commit('team/SET_CURRENT_TEAM', partnerTeam)
  const wrapper = shallowMount(SidebarOverlay, { localVue, store, stubs, mocks })
  const model = new Model(wrapper)
  expect(model.currentClients.length).toEqual(2)
})

it('assigns prefixed key to unselected team clients', () => {
  store.commit('team/SET_CURRENT_TEAM', partnerTeam)
  const wrapper = shallowMount(SidebarOverlay, { localVue, store, stubs, mocks })
  const model = new Model(wrapper)
  expect(model.currentClients[0].key.startsWith('client')).toBe(true)
})
