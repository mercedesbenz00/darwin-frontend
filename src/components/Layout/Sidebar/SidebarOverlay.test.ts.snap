// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div class="overlay">
  <div class="overlay-wrapper" style="--max-team-amount: 5; --el-height: 61px;">
    <team-item-stub team="[object Object]" class="overlay__team overlay__team--current"></team-item-stub>
    <team-item-stub team="[object Object]" class="overlay__team overlay__team--other"></team-item-stub>
    <team-item-stub team="[object Object]" class="overlay__team overlay__team--other"></team-item-stub>
    <team-item-stub team="[object Object]" class="overlay__team overlay__team--other-client"></team-item-stub>
    <team-item-stub team="[object Object]" class="overlay__team overlay__team--other-client"></team-item-stub>
  </div>
  <div class="overlay__actions">
    <settings-item-stub class="overlay__settings"></settings-item-stub>
    <feature-guard-stub feature="WORKER_MODE_TOGGLE">
      <worker-mode-item-stub class="overlay__worker-mode"></worker-mode-item-stub>
    </feature-guard-stub>
    <div class="overlay__logout">
      <power-icon-stub class="overlay__logout__icon"></power-icon-stub>
      <div class="overlay__logout__text">
        Log Out
      </div>
    </div>
  </div>
</div>
`;
